### ISSUE TITLE
Proposal: adopt rethinkdbdash as our official JS/node driver

### ISSUE THREAD

**Issue by danielmewes** [rethinkdb/rethinkdb#3771]
@neumino's rethinkdbdash has a lot of interesting features that our official JS driver doesn't currently provide.
Among them:
- Support for node streams (readable and writeable), would solve https://github.com/rethinkdb/rethinkdb/issues/3600 
- Written in pure JavaScript, making it easier for many developers to understand and contribute to https://github.com/rethinkdb/rethinkdb/issues/766
- A connection pool

There are a few things that we would need to change to be consistent with our other drivers:
- not automatically coerce streams to arrays
- think about whether we want to keep the connection pool (in its current form). See https://github.com/rethinkdb/rethinkdb/issues/281 (Probably we should just keep it. Even if our other drivers won't have one at first, it seems to be extremely useful in a node.js environment)
- rename the module

So we would probably have to fork it, but should make a serious attempt at diverging as little as possible to allow future patches to be shared between the projects.

(There might be some other differences to our official driver that I'm not aware of / forgot to mention.)

**Comment by deontologician** [comment]
Alternately, we might look at integrating parts of RethinkDB dash into the official driver, depending on how much is different. (i.e. copy the connection pool and stream implementations etc), and leave RethinkDB dash as a bleeding edge driver that can try out experimental ideas

**Comment by chrisvariety** [comment]
It's super silly and confusing to end users (like me :stuck_out_tongue_winking_eye:) to have two competing mostly official node.js rethinkdb drivers. (considering @neumino is a RethinkDB employee, I consider rethinkdbdash "mostly official"-- it's very different than someone random publishing another node driver.)

It's not like rethinkdbdash is just some sugar on top of the official driver, either -- it's a complete rewrite as far as I can tell, in a different language no less, which is kind of frightening.

**Comment by danielmewes** [comment]
Yeah I think just taking parts of it wouldn't quite work for the reasons @chrisvariety mentions.
(side note: @neumino is now at Google, but he's still maintaining and extending rethinkdbdash)

**Comment by thelinuxlich** [comment]
+1 for this, rethinkdbdash is awesome.

**Comment by marshall007** [comment]
I don't think this would inherently solve #3600. Even though @neumino recently introduced streaming support, the `cursor` object itself still doesn't implement the `Stream` interface.

I do like the idea of making it the official driver, though!

**Comment by AtnNn** [comment]
I think forking rethinkdb-dash seems like a bad idea. One of the reasons it is good is because @neumino maintains it and knows the ecosystem well and doesn't hesitate to break compatibility with the other drivers.

Another thing to consider is that it would need to be adapted to run in the data explorer.

**Comment by danielmewes** [comment]
@marshall007 as far as I understand implementing the `Stream` interface on the cursor directly would be risky because of the special meaning of `null` in Streams (as discussed in #3600).

@AtnNn It would definitely be nice if we could avoid forking, but since we want to do some things differently there might be no way around it.
We could consider just maintaining a patch file and applying it to upstream rethinkdbdash. That way we would be less likely to diverge too far.
Also good point about the Data Explorer, I forgot that in my list.

**Comment by marshall007** [comment]
It would probably be better if we could avoid porting `rethinkdbdash` to the browser. I'd imagine it would be much simpler if the browser sent the plain-text query to a separate (Node?) process which would evaluate the query and send the results back as JSON.

Something like that would allow all the HTTP stuff to be dropped from `rethinkdb` core and open up the possibility of supporting other languages in the Data Explorer in the future.

I'll probably take a stab at this and see how it goes if that seems reasonable to you guys.

**Comment by neumino** [comment]
A few thoughts:
- You can easily replace the official driver with rethinkdbdash:
https://github.com/neumino/rethinkdbdash#drop-in
In the end by setting the appropriate default, things will work out of the box like the official driver.
The differences after that are almost none. The only one I can think of is that fields that map to `undefined` are just drop in rethinkdbdash and that there is no option for that -- The official driver should probably do it too, see https://github.com/rethinkdb/rethinkdb/issues/2233

Then `rethinkdbdash` has more features like streams, writable streams, connection pool etc, but it still strictly implement all the ReQL methods like the official driver (`filter`, `update` etc.)

However, it does not support HTTP connections, which is probably a blocking point now. Nothing prevents HTTP connections to be supported as far as I know.

- I think it's still interesting to have a bleeding edge driver. I took me a few tries to implement writable/transform streams (this time 1.16.10 should be good), so it's still nice to have a playground for new stuff. For example, having `run({stream: true})` doesn't make that much sense, but I still did it because I didn't think enough. The official driver doesn't have to fall in this pitfall.

- @deontologician -- It's a decent effort to implement the connection pool. While I would love to say that I did it right on the first try, there were still a few bugs. If you add multiple servers on top of that, this gets tricky. The devil is in the detail :(

- @chrisvariety -- rethinkdbdash is still maintained, and I think I'm still doing a decent job. If I'm not mistaken 1.16 was released on the same day as the server :)

- It's probably a decent amount of work for @AtnNn and @larkost to switch driver. Backtraces are typically different in rethinkdbdash. Arity errors are returned by the driver if possible too.

- @marshall007 -- yes cursors still exist for two reasons:
  - support `null` values like @danielmewes mentionned
  - stay relatively close to the official driver

- Also on a side note, the python driver is going to nicer/more Python-ish way to consume feeds. I think it's fine to have drivers "drift" a bit when it comes to cursors, feeds etc. as long as the ReQL method themselves stay the same.

**Comment by joaojeronimo** [comment]
Massive +1 for this, especially because you advertise your coffeescript driver as a javascript driver (which is not a problem until someone decides to poke around).

Also maybe drop the bluebird dependency and use the 6to5 tool use standard ES6 promises ?

**Comment by thelinuxlich** [comment]
Don't do that, native promise is slower!

**Comment by danielmewes** [comment]
@thelinuxlich you mean rethinkdbdash's promises are slower than with our official driver? May I ask how exactly you are measuring them? There might be a way to fix that.

**Comment by danielmewes** [comment]
(never mind, I missed that you were talking about ES6 promises)

**Comment by stuartpb** [comment]
I'd be okay with this, so long as the paradigm of constructing abstract queries and running them against separate connections/pools is still possible:

```
var appconn, infoconn;

// two separate clusters/connections
r.connect({host: appRdbAddress}).then(conn => appconn = conn);
r.connect({host: infoRdbAddress}).then(conn => infoconn = conn);

// one common complex query
var activeAuthors = r.table('authors')
  .between(r.now().sub(archiveLimit),r.now(),{index:'last_login'})
  .orderBy(r.desc(function(subject){
    var staleness = subject('last_checked').sub(r.now());
    var age = subject('last_login').sub(r.now());
    return staleness.div(age);
  }))
  .limit(1)(0);

// query can be reused across connections
activeAuthors.run(appcon).then(notifyActiveAppAuthors);
activeAuthors.run(infocon).then(announceActiveInfoAuthors);
```

**Comment by neumino** [comment]
As mentionned in https://github.com/rethinkdb/rethinkdb/issues/4316, this is actually not possible.
While the connection are almost the same, they don't behave the same way because of the way queries are built.

**Comment by stuartpb** [comment]
I'm talking about the hypothetical scenario where both the connections and the queries are coming from the same `rethinkdbdash`-renamed-to-`rethinkdb` driver `r`.

**Comment by stuartpb** [comment]
To clarify, in addition to the stuff mentioned in https://github.com/rethinkdb/rethinkdb/issues/3771#issue-57514922 (ie. not coercing cursors to arrays), these are the properties I would want/need of a new `rethinkdbdash`-modeled `rethinkdb` JS driver (I'm not sure how much of this is currently true of rethinkdbdash):

*(For clarity, I will refer to the current module returned by `require('rethinkdb')` as `old_r` and the hypothetical new module that would more closely resemble `require('rethinkdbdash')` as `new_r`. Note that `new_r` refers to the `require('rethinkdbdash')` module object, and **not** an instance of *calling* `require('rethinkdbdash')()`.)*

- `new_r` contains all the methods and properties of `old_r`, particularly those related to query-building. Queries built with the root `new_r` are "connectionless" (like queries built with `old_r` are/were).
  - Calling `.then` or `.run()` with no arguments on a "connectionless" query is an error. Connectionless queries *must* be run with an explicitly-specified connection.
  - Connectionless queries may provide some kind of `.bind(conn)` method, which returns a clone of the query that will use the given connection implicitly when run with `.then` or paramless `.run()`.
- Calling `new_r(opts)` as a function synchronously returns an `old_r`-like instance (the way `require('rethinkdbdash')()` does now), one which can *also* be passed to a constructed query's `.run` as a connection (see #4315).
  - `new_r(opts)` takes the **exact** same set of options regarding the host (or hosts, for cluster pooling) to connect to as `new_r.connect(opts)`.
  - Queries constructed from a `new_r(opts)` instance can use `.then()` or `.run()` without a connection parameter, for an implicit `.run` using the established connection.
  - Running a query against a connection created with `new_r(opts)` (either implicitly **or** by using it as an explicit argument to a query's `.run()`) before the connection is established (ie. synchronously immediately after creation) buffers the query for execution after the connection is established (as I believe rethinkdbdash does currently, at least for implicit-connection runs - again see #4315).
  - Queries constructed from a `new_r(opts)` instance should still support running against an alternative connection provided as a parameter to `.run`.
- Calling `new_r.connect(opts)` provides the **exact** same value as `new_r(opts)` would (ie. with query-builder methods that create queries with implicit bound connections), but asynchronously (via callbacks/promises), and only after the connection is established. (This is mostly for backwards compatibility, although there are also paradigms where it makes sense to only use the connection once it's been established.)
  - The return value of `new_r(opts)` may do well to have some kind of equivalent method like `.onConnect` (or even just `.connect()`), to allow these same behaviors that should only be applied once the connection is established. (In the event that queries have already been buffered, the callbacks/promises registered via `.onConnect` should fire first, as this allows for things like initialization.)

**Comment by stuartpb** [comment]
Heck, I could probably implement an example of what I'm talking about on top of the existing `rethinkdb` driver, although it wouldn't have `rethinkdbdash`'s stream2 support.

**Comment by stuartpb** [comment]
Oh, I guess the `r(connection_opts)` syntax would conflict with the current handling of `r` as a synonym for `r.expr`, huh? That's a bummer. Maybe it could be deprecated, the context-instantiating version of `r` given a name like `r.connection` or `r.context`, and `r` itself changed to a synonym of `r.context` in the next major release?

**Comment by stuartpb** [comment]
That, or `r()`-constructed connection objects could be designed to pull double-duty, using their config object as an `r.expr` if called with an interface that only makes sense when used as `r.expr()`, and running a query when called in a way that would only work as `r.connection()`, just as an *ugly* compatibility hack until `r()`-as-expr's deprecation can fully take and be fully replaced. (The most frustrating thing would be that `r()`-as-either would have to defer the connection establishment to do this, in a way that `r.connection()` would not.)

**Comment by chrisfosterelli** [comment]
If this was done, I think it'd also be important to start having rethinkdbdash start following semantic versioning. Currently npm defaults to `^` for versions, which leaves almost everyone using rethinkdbdash vulnerable to random breakage.

Most people who use rethinkdbdash have gotten used to this, but if the official driver suddenly moved to being non-semver during the transition that would likely cause a lot of confusion.

**Comment by deontologician** [comment]
@chrisfosterelli I agree we should use semver, but for the record, the official driver doesn't use semver either (unfortunately)

**Comment by chrisfosterelli** [comment]
Ah, sorry my bad! Never noticed dependency issues so I assumed :)

**Comment by danielmewes** [comment]
@chrisfosterelli Also see https://github.com/rethinkdb/rethinkdb/issues/5625

**Comment by gabor-boros** [comment]
The official driver could be replaced by rethinkdb-ts instead of rethinkdash since it is maintained, still a drop-in replacement and up to date with the latest standards.
